File: diskcount/sources/pricepertb.py
from __future__ import annotations

from decimal import Decimal

import httpx
from bs4 import BeautifulSoup

from diskcount.domain import Deal
from diskcount.parsing import (
    extract_asin,
    normalize_condition,
    normalize_drive_category,
    normalize_interfaces,
    normalize_media_type,
    parse_capacity_tb,
    parse_price_eur,
)


class PricePerTBSource:
    name = "pricepertb"

    def __init__(self, urls: list[str]) -> None:
        self.urls = urls

    async def fetch(self, client: httpx.AsyncClient) -> list[Deal]:
        deals: list[Deal] = []
        for url in self.urls:
            response = await client.get(url)
            response.raise_for_status()
            deals.extend(parse_pricepertb_html(response.text))
        return deals


def _cell_text(cell) -> str:
    return cell.get_text(" ", strip=True)


def parse_pricepertb_html(html: str) -> list[Deal]:
    soup = BeautifulSoup(html, "html.parser")
    deals: list[Deal] = []

    for row in soup.select("tr"):
        cells = row.find_all(["td", "th"])
        if len(cells) < 8:
            continue
        texts = [_cell_text(cell) for cell in cells]
        if any("Price" in text or "Prix" in text for text in texts[:2]):
            continue

        if len(cells) >= 9:
            price_per_tb_text = texts[1]
            price_text = texts[2]
            capacity_text = texts[3]
            warranty_text = texts[4]
            form_factor_text = texts[5]
            technology_text = texts[6]
            condition_text = texts[7]
            link_cell = cells[8]
        else:
            price_per_tb_text = texts[0]
            price_text = texts[1]
            capacity_text = texts[2]
            warranty_text = texts[3]
            form_factor_text = texts[4]
            technology_text = texts[5]
            condition_text = texts[6]
            link_cell = cells[7]

        price_per_tb = parse_price_eur(price_per_tb_text)
        price_eur = parse_price_eur(price_text)
        capacity_tb = parse_capacity_tb(capacity_text)
        if price_per_tb is None or price_eur is None or capacity_tb is None or capacity_tb <= 0:
            continue

        media_type = normalize_media_type(" ".join([technology_text, form_factor_text]))
        if media_type not in ("rotational", "solid_state"):
            continue

        link = link_cell.find("a", href=True)
        if not link:
            continue
        url = str(link["href"])
        title = link.get_text(" ", strip=True)
        category_text = " ".join([form_factor_text, technology_text, title])
        condition = normalize_condition(condition_text)
        drive_category = normalize_drive_category(category_text, media_type)
        interfaces = normalize_interfaces(category_text)

        deals.append(
            Deal(
                source=PricePerTBSource.name,
                external_id=extract_asin(url),
                title=title,
                url=url,
                price_eur=price_eur.quantize(Decimal("0.01")),
                price_per_tb=price_per_tb.quantize(Decimal("0.01")),
                capacity_tb=capacity_tb,
                condition=condition,
                media_type=media_type,
                form_factor=form_factor_text or None,
                technology=technology_text or None,
                drive_category=drive_category,
                interfaces=interfaces,
                raw={"warranty": warranty_text},
            )
        )

    return deals

File: internal/sources/pricepertb.go
package sources

import (
	"context"
	"net/url"
	"strings"

	"github.com/Balrog57/DiskCount/internal/domain"
	"github.com/Balrog57/DiskCount/internal/parsing"
	"github.com/Balrog57/DiskCount/internal/scraper"
	"github.com/PuerkitoBio/goquery"
)

func init() {
	Register(func(r *Registry) Source {
		cfg := r.Config()
		if len(cfg.PricePerTBURLs) == 0 {
			return nil
		}
		return &PricePerTB{http: r.HTTP(), byparr: r.Byparr(), urls: cfg.PricePerTBURLs, useFB: cfg.HeadlessFallback}
	})
}

type PricePerTB struct {
	http   scraper.Fetcher
	byparr *scraper.ByparrClient
	urls   []string
	useFB  bool
}

func (s *PricePerTB) Name() string { return "pricepertb" }

func (s *PricePerTB) Info() SourceInfo {
	return SourceInfo{
		Name:        "pricepertb",
		Description: "Tableau HTML pricepertb.com (€/To, multi-pays)",
		Categories:  []string{"scraping"},
		Requires:    []string{"PRICEPERTB_URLS"},
		Version:     "1",
	}
}

func (s *PricePerTB) Fetch(ctx context.Context) ([]domain.Deal, error) {
	res := fetchMultiURL(ctx, s.Name(), s.http, s.byparr, s.urls, s.useFB, parsePTB)
	if err := res.asTransientError(s.Name()); err != nil {
		return nil, err
	}
	return res.deals, nil
}

func parsePTB(html, baseURL string) []domain.Deal {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil
	}
	var deals []domain.Deal
	doc.Find("tr").Each(func(_ int, row *goquery.Selection) {
		cells := row.Find("td")
		if cells.Length() < 3 {
			return
		}
		texts := make([]string, cells.Length())
		cells.Each(func(i int, c *goquery.Selection) { texts[i] = strings.TrimSpace(c.Text()) })
		if strings.Contains(strings.ToLower(texts[0]), "price") {
			return
		}

		priceEUR, err := parseFloatClean(textAt(texts, 2))
		if err != nil || priceEUR <= 0 {
			return
		}
		pricePerTB, err := parseFloatClean(strings.TrimSpace(row.Find("td.price-per-tb").First().Text()))
		if err != nil || pricePerTB <= 0 {
			return
		}
		capText := textAt(texts, 3)
		var tb float64
		if attr, ok := row.Attr("data-capacity"); ok {
			if gb, e := parseFloatClean(attr); e == nil && gb > 0 {
				tb = gb / 1000
			}
		}
		if tb <= 0 {
			if parsed, e := parsing.ParseCapacityTB(capText); e == nil && parsed > 0 {
				tb = parsed
			}
		}
		if tb <= 0 {
			if parts := strings.Fields(capText); len(parts) > 0 {
				if v, e := parseFloatClean(parts[0]); e == nil && v > 0 {
					tb = v
					if len(parts) > 1 && strings.EqualFold(parts[1], "Go") {
						tb /= 1000
					}
				}
			}
		}
		if tb <= 0 {
			return
		}
		link := row.Find("a").First()
		href, _ := link.Attr("href")
		href = absolutizeURL(baseURL, strings.TrimSpace(href))
		if !isAmazonURL(href) {
			return
		}
		title := strings.TrimSpace(link.Text())
		if title == "" && len(texts) > 2 {
			title = strings.TrimSpace(texts[2])
		}
		if title == "" || href == "" {
			return
		}

		formFactor := textAt(texts, 5)
		technology := textAt(texts, 6)
		classText := strings.Join([]string{title, formFactor, technology}, " ")
		media := normalMedia(classText)
		if media == nil {
			return
		}
		dc := parsing.NormalizeDriveCategory(classText, media)
		ifaces := parsing.NormalizeInterfaces(classText)
		if len(ifaces) == 0 && dc != nil {
			ifaces = parsing.DefaultInterfacesForCategory(*dc)
		}

		cond := domain.ConditionNew
		condText := strings.TrimSpace(textAt(texts, 7))
		if attr, ok := row.Attr("data-condition"); ok && strings.TrimSpace(attr) != "" {
			condText = attr
		}
		if strings.Contains(strings.ToLower(condText), "used") || strings.Contains(strings.ToLower(condText), "occasion") {
			cond = domain.ConditionUsed
		}
		deals = append(deals, domain.Deal{
			Source: "pricepertb", Title: title, URL: href,
			PriceEUR: round2(priceEUR), PricePerTB: round2(pricePerTB), CapacityTB: round2(tb),
			Condition: &cond, MediaType: media,
			ExternalID:    parsing.ExtractASIN(href),
			SKU:           amazonSKU("", parsing.ExtractASIN(href)),
			ImageURL:      amazonImageFromASIN(parsing.ExtractASIN(href)),
			DriveCategory: dc, Interfaces: ifaces,
			FormFactor: strPtr(formFactor), Technology: strPtr(technology),
			ObservedAt: domain.UTCNow(),
		})
	})
	return deals
}

func textAt(texts []string, i int) string {
	if i < 0 || i >= len(texts) {
		return ""
	}
	return strings.TrimSpace(texts[i])
}

func absolutizeURL(baseURL, raw string) string {
	if raw == "" {
		return ""
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	if parsed.IsAbs() {
		return parsed.String()
	}
	base, err := url.Parse(baseURL)
	if err != nil {
		return raw
	}
	return base.ResolveReference(parsed).String()
}
